File: MPI/03.Ring/ring-one-sided.cpp
#include <mpi.h>
#include <vector>

void copyTo(std::vector<int> &dst, std::vector<int> &src){
  for(int i = 0; i < src.size(); i++){
    dst.at(i) = src[i];
  }
}

int main(int argc, char* argv[]){
  int my_rank, num_procs, sum = 0;

  MPI_Init(&argc, &argv);
  MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);
  MPI_Comm_size(MPI_COMM_WORLD, &num_procs);

  std::vector<int> rcv_buf = { 0 };
  std::vector<int> snd_buf = { my_rank };
  int right = (my_rank+1) % num_procs;
  int left = (my_rank-1+num_procs) % num_procs;
  MPI_Win win;

  MPI_Win_create(&snd_buf[0], sizeof(snd_buf), sizeof(MPI_INT), MPI_INFO_NULL, MPI_COMM_WORLD, &win);

  for(int i = 0; i < num_procs; i++){
    MPI_Win_fence(0, win);
    MPI_Get(&rcv_buf[0], 1, MPI_INT, left, 0, 1, MPI_INT, win);
    MPI_Win_fence(0, win);
    copyTo(snd_buf, rcv_buf);
    sum += rcv_buf[0];
  }

  MPI_Win_free(&win);

  printf("PE%d:\tSum = %d\n", my_rank, sum);

  MPI_Finalize();
}

File: MPI/03.Ring/ring-one-sided.py
from mpi4py import MPI
import numpy as np

rcv_buf = np.empty((), dtype=np.intc)

comm_world = MPI.COMM_WORLD
my_rank = comm_world.Get_rank()
size = comm_world.Get_size()

right = (my_rank+1) % size
left  = (my_rank-1+size) % size

sum = 0
snd_buf = np.array(my_rank, dtype=np.intc)

win = MPI.Win.Create(snd_buf, comm=comm_world)

for i in range(size):
  win.Fence()
  win.Get([rcv_buf, MPI.INT], left)
  win.Fence()
  sum += rcv_buf
  np.copyto(snd_buf, rcv_buf)

win.Free()

print(f"PE{my_rank}:\tSum = {sum}")